File: patches/fix_cowork_sandbox_refs.py
#!/usr/bin/env python3
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: python
"""
Rewrite cowork sandbox/VM references for runtime-selected backend.

Upstream tool descriptions + system prompts tell the model it runs in a
"lightweight Linux VM (Ubuntu 22)" or "isolated Linux environment". That is
accurate when the kvm backend is running but wrong on native — the native
backend executes Claude Code directly on the host.

Each patch wraps the affected text in a JS ternary keyed off
`globalThis.__coworkKvmMode`, which is set by `fix_cowork_linux.py`'s
preamble based on `COWORK_VM_BACKEND` / socket auto-detect:

    globalThis.__coworkKvmMode ? "<original VM-accurate text>"
                               : "<host-accurate rewrite>"

The injection style depends on the JS string-literal kind that surrounds
the target text:

  "…target…"        → close the dquote, concat a ternary, reopen:
                      "…" + (globalThis.__coworkKvmMode?"orig":"new") + "…"
  `…target…`        → template interpolation:
                      `…${globalThis.__coworkKvmMode?"orig":"new"}…`

Patches always run on clean, freshly-extracted bundles — no "already
patched" fast path; a missing anchor is a hard failure.

Patches:
  A) Bash tool description: "isolated Linux workspace" vs "host Linux system"
  B) Cowork identity prompt: "lightweight Linux VM" vs "directly on the host"
  C) Computer use explanation: "lightweight Linux VM (Ubuntu 22)" vs host
  D) System prompt + error messages: "isolated Linux environment" vs "host"

Usage: python3 fix_cowork_sandbox_refs.py <path_to_index.js>
"""

import sys
import os
import re


EXPECTED_PATCHES = 4


def _classify_context(content: bytes, pos: int) -> str:
    """Return 'backtick' or 'dquote' — whichever delimiter is nearest
    before `pos` is the enclosing literal kind. Used to pick the right
    injection style."""
    last_bt = content.rfind(b"`", 0, pos)
    last_dq = content.rfind(b'"', 0, pos)
    if last_bt > last_dq:
        return "backtick"
    return "dquote"


def _inject_ternary(context: str, orig: bytes, new: bytes) -> bytes:
    """Produce the bytes to replace `orig` with so the bundle evaluates to
    either `orig` or `new` at runtime depending on globalThis.__coworkKvmMode.

    Dquote context: `"…" + ternary + "…"` style.
    Backtick context: `…${ternary}…` style (raw `${}` — no extra quotes).
    """
    ternary = b'globalThis.__coworkKvmMode?"' + orig + b'":"' + new + b'"'
    if context == "backtick":
        return b"${" + ternary + b"}"
    return b'"+(' + ternary + b')+"'


def _replace_substring_context_aware(content: bytes, orig: bytes, new: bytes) -> tuple[bytes, int]:
    """Replace every occurrence of `orig` in `content` with a runtime
    ternary whose form depends on the enclosing string-literal kind.

    Scans left-to-right so each replacement's context is computed against
    the still-pristine prefix — fresh replacements don't confuse later
    context lookups (a backtick/dquote we insert lives downstream only)."""
    result = bytearray()
    cursor = 0
    count = 0
    while True:
        pos = content.find(orig, cursor)
        if pos < 0:
            result.extend(content[cursor:])
            break
        context = _classify_context(content, pos)
        result.extend(content[cursor:pos])
        result.extend(_inject_ternary(context, orig, new))
        cursor = pos + len(orig)
        count += 1
    return bytes(result), count


def patch_sandbox_refs(filepath):
    """Rewrite sandbox/VM references with runtime-selected text."""

    print("=== Patch: fix_cowork_sandbox_refs ===")
    print(f"  Target: {filepath}")

    if not os.path.exists(filepath):
        print(f"  [FAIL] File not found: {filepath}")
        return False

    with open(filepath, "rb") as f:
        content = f.read()

    original_content = content
    patches_applied = 0

    # ── Patch A: Bash tool description ───────────────────────────────
    #
    # Upstream: "Run a shell command in the session's isolated Linux
    #           workspace. Your connected folders are mounted under
    #           /sessions/" + t.vmProcessName + "/mnt/ — …"
    #
    # Both halves are plain "…" literals so use dquote-style injection
    # directly: wrap each half in a ternary. The dynamic concat (+expr+)
    # stays intact.
    pattern_a = re.compile(
        rb'("Run a shell command in the session\'s isolated Linux workspace\.'
        rb'[^"]*?/sessions/")'     # first half literal
        rb"(\+[\w$.]+\+)"          # dynamic concat (+ t.vmProcessName +)
        rb'("/mnt/[^"]*?")'        # second half literal
    )

    native_half1 = (
        b'"Run a shell command on the host Linux system.'
        b" There is no VM or sandbox \\u2014 commands execute directly"
        b" on the user\\u2019s computer."
        b" Each bash call is independent (no cwd/env carryover)."
        b' Use absolute paths."'
    )
    native_half2 = b'""'

    def repl_a(m):
        orig_half1 = m.group(1)
        concat = m.group(2)
        orig_half2 = m.group(3)
        return (
            b"(globalThis.__coworkKvmMode?" + orig_half1 + b":" + native_half1 + b")"
            + concat
            + b"(globalThis.__coworkKvmMode?" + orig_half2 + b":" + native_half2 + b")"
        )

    content, count_a = pattern_a.subn(repl_a, content, count=1)
    if count_a == 1:
        print("  [OK] A bash tool description: wrapped in runtime ternary")
        patches_applied += 1
    else:
        print("  [FAIL] A bash tool description: pattern not found")

    # ── Patch B: Cowork identity system prompt ───────────────────────
    # Embedded in a template literal (NMr = `…`) so injection uses `${…}`.
    orig_b = b"Claude runs in a lightweight Linux VM on the user's computer, which provides a secure sandbox for executing code while allowing controlled access to a workspace folder."
    new_b = b"Claude runs directly on the user's Linux computer with full access to the local filesystem and installed tools. There is no VM or sandbox."

    content, count_b = _replace_substring_context_aware(content, orig_b, new_b)
    if count_b >= 1:
        print(f"  [OK] B cowork identity prompt: wrapped {count_b} occurrence(s)")
        patches_applied += 1
    else:
        print("  [FAIL] B cowork identity prompt: pattern not found")

    # ── Patch C: Computer use high-level explanation ─────────────────
    orig_c = b"Claude runs in a lightweight Linux VM (Ubuntu 22) on the user's computer. This VM provides a secure sandbox for executing code while allowing controlled access to user files."
    new_c = b"Claude runs directly on the user's Linux computer. Commands execute on the host system with full access to local files and tools. There is no VM or sandbox."

    content, count_c = _replace_substring_context_aware(content, orig_c, new_c)
    if count_c >= 1:
        print(f"  [OK] C computer use explanation: wrapped {count_c} occurrence(s)")
        patches_applied += 1
    else:
        print("  [FAIL] C computer use explanation: pattern not found")

    # ── Patch D: "isolated Linux environment" variants ──────────────
    #
    # Two substring forms; each can appear in either a "…" string literal
    # (error messages) or a `…` template literal (system prompt). The
    # context-aware helper picks the right injection style per occurrence.
    variants = [
        (b"The isolated Linux environment", b"The host Linux environment"),
        (b"an isolated Linux environment", b"the host Linux environment"),
    ]

    total_d = 0
    for orig_sub, new_sub in variants:
        content, n = _replace_substring_context_aware(content, orig_sub, new_sub)
        total_d += n

    if total_d >= 1:
        print(f"  [OK] D isolated Linux environment: wrapped {total_d} occurrence(s)")
        patches_applied += 1
    else:
        print("  [FAIL] D isolated Linux environment: pattern not found")

    # ── Checks ───────────────────────────────────────────────────────
    if patches_applied < EXPECTED_PATCHES:
        print(f"  [FAIL] Only {patches_applied}/{EXPECTED_PATCHES} patches applied")
        return False

    original_delta = original_content.count(b"{") - original_content.count(b"}")
    patched_delta = content.count(b"{") - content.count(b"}")
    if original_delta != patched_delta:
        diff = patched_delta - original_delta
        print(f"  [FAIL] Patch introduced brace imbalance: {diff:+d} unmatched braces")
        return False

    with open(filepath, "wb") as f:
        f.write(content)
    print(f"  [PASS] All {patches_applied} sandbox/VM references wrapped for runtime selection")
    return True


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <path_to_index.js>")
        sys.exit(1)

    success = patch_sandbox_refs(sys.argv[1])
    sys.exit(0 if success else 1)

File: patches/fix_cowork_sandbox_refs.nim
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: nim
#
# Nim port of fix_cowork_sandbox_refs.py — produces byte-identical output.
#
# 4 sub-patches, each wraps a sandbox/VM-accurate phrase in a runtime
# three-way ternary keyed on globalThis.__coworkKvmMode and
# globalThis.__coworkSandboxMode (both set by fix_cowork_linux):
#
#   __coworkKvmMode      → original kvm/VM phrasing
#   __coworkSandboxMode  → bwrap sandbox phrasing
#   else                 → native (host, no isolation) phrasing
#
# Injection style depends on the enclosing JS string literal:
#   "…target…"  → close dquote + concat + reopen:
#                 "…"+(globalThis.__coworkKvmMode?"kvm":
#                       globalThis.__coworkSandboxMode?"sandbox":"native")+"…"
#   `…target…`  → template interpolation:
#                 `…${globalThis.__coworkKvmMode?"kvm":
#                       globalThis.__coworkSandboxMode?"sandbox":"native"}…`
#
# No "already patched" fast path — missing anchors are hard failures.

import std/[os, strformat, strutils, options]
import std/nre

proc replaceFirstRe(
    content: var string, pattern: Regex, subFn: proc(m: RegexMatch): string
): int =
  let maybe = content.find(pattern)
  if maybe.isNone:
    return 0
  let m = maybe.get()
  let bounds = m.matchBounds
  content = content[0 ..< bounds.a] & subFn(m) & content[bounds.b + 1 .. ^1]
  return 1

proc classifyContext(content: string, pos: int): string =
  ## Whichever of " / ` is nearer before `pos` wins. Matches Python's
  ## rfind(b"`", 0, pos) / rfind(b'"', 0, pos) comparison: lastBt > lastDq
  ## means the backtick is closer to pos.
  let lastBt = content.rfind("`", last = pos - 1)
  let lastDq = content.rfind("\"", last = pos - 1)
  if lastBt > lastDq: "backtick" else: "dquote"

proc injectTernary(context, kvmStr, sandboxStr, nativeStr: string): string =
  let ternary =
    "globalThis.__coworkKvmMode?\"" & kvmStr &
    "\":globalThis.__coworkSandboxMode?\"" & sandboxStr &
    "\":\"" & nativeStr & "\""
  if context == "backtick":
    "${" & ternary & "}"
  else:
    "\"+(" & ternary & ")+\""

proc replaceSubstringContextAware(
    content: var string, kvmStr, sandboxStr, nativeStr: string
): int =
  ## Walk left-to-right, replace each occurrence of kvmStr (the phrasing
  ## present in the upstream bundle) with a context-aware three-way
  ## ternary. Each replacement's context is classified against the prefix
  ## of the pre-edit string so fresh quotes/backticks we inject don't
  ## pollute later context lookups.
  var buf = newStringOfCap(content.len + 256)
  var cursor = 0
  var count = 0
  while true:
    let pos = content.find(kvmStr, cursor)
    if pos < 0:
      buf.add content[cursor .. ^1]
      break
    let context = classifyContext(content, pos)
    buf.add content[cursor ..< pos]
    buf.add injectTernary(context, kvmStr, sandboxStr, nativeStr)
    cursor = pos + kvmStr.len
    inc count
  content = buf
  return count

proc apply*(input: string): string =
  var content = input
  let original = input
  var patchesApplied = 0
  const EXPECTED_PATCHES = 8

  # ── Patch A: Bash tool description ─────────────────────────────
  # Upstream v1.7196.1 collapsed the prior three-piece concat
  #   "...sessions/" + VAR + "/mnt/..."
  # into a single literal that uses "<session>" as an in-string placeholder.
  # Match the whole literal and wrap it in the runtime ternary.
  block:
    let patNew =
      re"""("Run a shell command in the session's isolated Linux workspace\.[^"]*?wait a few seconds and retry\.")"""
    let patOld =
      re"""("Run a shell command in the session's isolated Linux workspace\.[^"]*?/sessions/")(\+[\w$.]+\+)("/mnt/[^"]*?")"""

    # Native-mode rewrite (byte-for-byte identical to the Python source).
    const nativeHalf1 =
      "\"Run a shell command on the host Linux system." &
      " There is no VM or sandbox \\u2014 commands execute directly" &
      " on the user\\u2019s computer." &
      " Each bash call is independent (no cwd/env carryover)." & " Use absolute paths.\""

    # Sandbox-mode rewrite — bwrap fallback backend.
    const sandboxHalf1 =
      "\"Run a shell command inside a bubblewrap sandbox on the user\\u2019s Linux machine." &
      " Writes are confined to the user-selected workspace and a session outputs" &
      " scratch directory; the host filesystem is otherwise read-only or hidden." &
      " Network egress is filtered through an allowlist proxy." &
      " Each bash call is independent (no cwd/env carryover)." & " Use absolute paths.\""

    var n = replaceFirstRe(
      content,
      patNew,
      proc(m: RegexMatch): string =
        let origStr = m.captures[0]
        "(globalThis.__coworkKvmMode?" & origStr & ":" &
          "globalThis.__coworkSandboxMode?" & sandboxHalf1 & ":" &
          nativeHalf1 & ")",
    )
    if n == 0:
      # Fallback to the pre-v1.7196.1 three-piece concat pattern.
      n = replaceFirstRe(
        content,
        patOld,
        proc(m: RegexMatch): string =
          let origHalf1 = m.captures[0]
          let concat = m.captures[1]
          let origHalf2 = m.captures[2]
          "(globalThis.__coworkKvmMode?" & origHalf1 & concat & origHalf2 & ":" &
            "globalThis.__coworkSandboxMode?" & sandboxHalf1 & ":" &
            nativeHalf1 & ")",
      )
    if n == 1:
      echo "  [OK] A bash tool description: wrapped in three-way runtime ternary"
      inc patchesApplied
    else:
      echo "  [FAIL] A bash tool description: pattern not found"

  # ── Patch B: Cowork identity system prompt ─────────────────────
  block:
    const kvmB =
      "Claude runs in a lightweight Linux VM on the user's computer, which provides a secure sandbox for executing code while allowing controlled access to a workspace folder."
    const sandboxB =
      "Claude runs inside a bubblewrap sandbox on the user's Linux computer. The sandbox restricts writes to the user-selected workspace and a session outputs directory, exposes most of the host filesystem read-only, and routes network traffic through an allowlist proxy."
    const nativeB =
      "Claude runs directly on the user's Linux computer with full access to the local filesystem and installed tools. There is no VM or sandbox."

    let n = replaceSubstringContextAware(content, kvmB, sandboxB, nativeB)
    if n >= 1:
      echo &"  [OK] B cowork identity prompt: wrapped {n} occurrence(s)"
      inc patchesApplied
    else:
      echo "  [FAIL] B cowork identity prompt: pattern not found"

  # ── Patch C: Computer use high-level explanation ───────────────
  block:
    const kvmC =
      "Claude runs in a lightweight Linux VM (Ubuntu 22) on the user's computer. This VM provides a secure sandbox for executing code while allowing controlled access to user files."
    const sandboxC =
      "Claude runs inside a bubblewrap sandbox on the user's Linux computer. Commands execute on the host as the user but with restricted filesystem scope (writes confined to the user-selected workspace and outputs dir, host otherwise read-only) and an allowlisted network proxy. There is no VM but there is a sandbox."
    const nativeC =
      "Claude runs directly on the user's Linux computer. Commands execute on the host system with full access to local files and tools. There is no VM or sandbox."

    let n = replaceSubstringContextAware(content, kvmC, sandboxC, nativeC)
    if n >= 1:
      echo &"  [OK] C computer use explanation: wrapped {n} occurrence(s)"
      inc patchesApplied
    else:
      echo "  [FAIL] C computer use explanation: pattern not found"

  # ── Patch D: "isolated Linux environment" variants ────────────
  block:
    let variants = [
      ("The isolated Linux environment", "The sandboxed Linux environment", "The host Linux environment"),
      ("an isolated Linux environment", "a sandboxed Linux environment", "the host Linux environment"),
    ]
    var totalD = 0
    for (kvmSub, sandboxSub, nativeSub) in variants:
      totalD += replaceSubstringContextAware(content, kvmSub, sandboxSub, nativeSub)
    if totalD >= 1:
      echo &"  [OK] D isolated Linux environment: wrapped {totalD} occurrence(s)"
      inc patchesApplied
    else:
      echo "  [FAIL] D isolated Linux environment: pattern not found"

  # ── Patch E: "Files you create in the sandbox" sentence ───────
  # Two occurrences in the CU section, one per template-literal var
  # name (the workspace path interpolation). The minified name shifts
  # across upstream releases (v1.6608: ${AA}; v1.7196.1: ${eA}; the
  # first variant has stayed at ${e}). Match with [\w$]+ so this
  # survives future renames. In native cowork mode the whole sentence
  # is misleading — there is no sandbox, files Claude creates are
  # directly visible to host desktop apps. Omit the sentence in
  # native; keep upstream in kvm/sandbox.
  block:
    let patE =
      re"""Files you create in the sandbox \(under \\`\$\{[\w$]+\}\\` or \\`/tmp\\`\) do NOT exist on the user's machine\. If you put a command or file path in the user's clipboard, or type into one of their apps, the path must exist on THEIR computer — not a sandbox path they can't reach\."""
    var hitsE = 0
    content = content.replace(
      patE,
      proc(m: RegexMatch): string =
        inc hitsE
        "${(globalThis.__coworkKvmMode||globalThis.__coworkSandboxMode)?`" &
          m.match & "`:\"\"}",
    )
    if hitsE >= 2:
      echo &"  [OK] E sandbox-file warning: omitted in native ({hitsE} occurrences)"
      inc patchesApplied
    else:
      echo &"  [FAIL] E sandbox-file warning: expected 2, found {hitsE}"

  # ── Patch F: Shell access — "Paths in bash differ ... translate." ──
  # The block interpolates 3 template-vars on the mapping line and 2
  # path-vars on the example line. All five names are minified and
  # have shifted across releases (v1.6608: ${Ae}${EA}${W} / ${dA},${nA};
  # v1.7196.1: ${uA}${W}${IA} / ${nA},${lA}). Match each with [\w$]+
  # so renames don't break the patch.
  # In native mode bash and the file tools share the same filesystem;
  # the path-mapping block (and the "translate" instruction) is wrong
  # and has actually misled the model in practice. Replace with a
  # single line clarifying that no translation is needed.
  block:
    let patF =
      re"""Paths in bash differ from what file tools \(Read/Write/Edit\) see:\n\$\{[\w$]+\}\$\{[\w$]+\}\$\{[\w$]+\}\n\nSo a file you Read at \$\{[\w$]+\}/foo\.txt is reached in bash at \$\{[\w$]+\}/foo\.txt — use the mapping above to translate\."""
    var hitsF = 0
    content = content.replace(
      patF,
      proc(m: RegexMatch): string =
        inc hitsF
        "${(globalThis.__coworkKvmMode||globalThis.__coworkSandboxMode)?`" &
          m.match &
          "`:\"Bash and the file tools (Read/Write/Edit) see the same filesystem \\u2014 no path translation needed; use absolute paths directly.\"}",
    )
    if hitsF == 1:
      echo "  [OK] F shell-access path mapping: native gets same-fs note"
      inc patchesApplied
    else:
      echo &"  [FAIL] F shell-access path mapping: expected 1, found {hitsF}"

  # ── Patch G: Skill scripts via VM path ─────────────────────────
  # The "VM path above" wording only makes sense when there IS a
  # separate VM-side path. In native mode the file-tool path and the
  # bash path are the same, so this sentence is wrong (and the "above"
  # reference dangles since Patch F removes the mapping). Suppress in
  # native cowork mode.
  # The gating var was `M` in earlier releases and `N` in v1.7196.1+;
  # match with [\w$]+ to ride out future renames.
  block:
    let patG =
      re"""\+\(([\w$]+)\?" Skill scripts can be run via bash using the VM path above\."\:""\)\+"""
    let n = replaceFirstRe(
      content,
      patG,
      proc(m: RegexMatch): string =
        let gateVar = m.captures[0]
        "+((" & gateVar &
          "&&(globalThis.__coworkKvmMode||globalThis.__coworkSandboxMode))?\"" &
          " Skill scripts can be run via bash using the VM path above.\":\"\")+",
    )
    if n == 1:
      echo "  [OK] G skill-scripts VM path: cowork-only"
      inc patchesApplied
    else:
      echo "  [FAIL] G skill-scripts VM path: anchor not found"

  # ── Patch H: "Linux environment boots in the background" ───────
  # No Linux environment boots in native cowork mode — the shell is
  # just the user's host. Omit the sentence (and its leading blank
  # line) in native; keep upstream in kvm/sandbox.
  block:
    const anchor =
      "\n\nThe Linux environment boots in the background. If bash returns \"Workspace still starting\", wait a few seconds and retry."
    if content.contains(anchor):
      let repl =
        "${(globalThis.__coworkKvmMode||globalThis.__coworkSandboxMode)?`" &
        anchor & "`:\"\"}"
      content = content.replace(anchor, repl)
      echo "  [OK] H workspace-starting notice: cowork-only"
      inc patchesApplied
    else:
      echo "  [FAIL] H workspace-starting notice: anchor not found"

  # ── Checks ────────────────────────────────────────────────────
  if patchesApplied < EXPECTED_PATCHES:
    raise newException(
      ValueError, &"Only {patchesApplied}/{EXPECTED_PATCHES} patches applied"
    )

  let originalDelta = original.count('{') - original.count('}')
  let patchedDelta = content.count('{') - content.count('}')
  if originalDelta != patchedDelta:
    let diff = patchedDelta - originalDelta
    raise newException(
      ValueError, &"Patch introduced brace imbalance: {diff:+d} unmatched braces"
    )

  echo &"  [PASS] All {patchesApplied} sandbox/VM references wrapped for runtime selection"
  result = content

when isMainModule:
  let file = paramStr(1)
  echo "=== Patch: fix_cowork_sandbox_refs ==="
  echo &"  Target: {file}"
  let input = readFile(file)
  let output = apply(input)
  if output != input:
    writeFile(file, output)
